src/order/jordan_hoelder.lean

```diff
@@ -0,0 +1,803 @@
+--     rw [mem_def, mem_def, h]
+--   end }
+
+lemma length_pos_of_mem_ne {s : composition_series X}
+  {x y : X} (hx : x ∈ s) (hy : y ∈ s) (hxy : x ≠ y) :
+  0 < s.length :=
+let ⟨i, hi⟩ := hx, ⟨j, hj⟩ := hy in
+have hij : i ≠ j, from mt s.inj.2 $ λ h, hxy (hi ▸ hj ▸ h),
+hij.lt_or_lt.elim
+  (λ hij, (lt_of_le_of_lt (zero_le i)
+    (lt_of_lt_of_le hij (nat.le_of_lt_succ j.2))))
+  (λ hji, (lt_of_le_of_lt (zero_le j)
+    (lt_of_lt_of_le hji (nat.le_of_lt_succ i.2))))
+
+lemma forall_mem_eq_of_length_eq_zero {s : composition_series X}
+  (hs : s.length = 0) {x y} (hx : x ∈ s) (hy : y ∈ s) : x = y :=
+by_contradiction (λ hxy, pos_iff_ne_zero.1 (length_pos_of_mem_ne hx hy hxy) hs)
+
+/-- Remove the largest element from a `composition_series`. If the series `s`
+has length zero, then `s.erase_top = s` -/
+@[simps] def erase_top (s : composition_series X) : composition_series X :=
+{ length := s.length - 1,
+  series := λ i, s ⟨i, lt_of_lt_of_le i.2 (nat.succ_le_succ (nat.sub_le_self _ _))⟩,
+  step' := λ i, begin
+    have := s.step ⟨i, lt_of_lt_of_le i.2 (nat.sub_le_self _ _)⟩,
+    cases i,
+    exact this
+  end }
+
+lemma top_erase_top (s : composition_series X) :
+  s.erase_top.top = s ⟨s.length - 1, lt_of_le_of_lt (nat.sub_le_self _ _) (nat.lt_succ_self _)⟩ :=
+show s _ = s _, from congr_arg s
+begin
+  ext,
+  simp only [erase_top_length, fin.coe_last, fin.coe_cast_succ, fin.coe_of_nat_eq_mod,
+    fin.coe_mk, coe_coe]
+end
+
+lemma erase_top_top_le (s : composition_series X) : s.erase_top.top ≤ s.top :=
+by simp [erase_top, top, s.strict_mono.le_iff_le, fin.le_iff_coe_le_coe, nat.sub_le_self]
+
+@[simp] lemma bot_erase_top (s : composition_series X) : s.erase_top.bot = s.bot := rfl
+
+lemma mem_erase_top_of_ne_of_mem {s : composition_series X} {x : X}
+  (hx : x ≠ s.top) (hxs : x ∈ s) : x ∈ s.erase_top :=
+begin
+  { rcases hxs with ⟨i, rfl⟩,
+    have hi : (i : ℕ) < (s.length - 1).succ,
+    { conv_rhs { rw [← nat.succ_sub (length_pos_of_mem_ne ⟨i, rfl⟩ s.top_mem hx),
+        nat.succ_sub_one] },
+      exact lt_of_le_of_ne
+        (nat.le_of_lt_succ i.2)
+        (by simpa [top, s.inj, fin.ext_iff] using hx) },
+    refine ⟨i.cast_succ, _⟩,
+    simp [fin.ext_iff, nat.mod_eq_of_lt hi] }
+end
+
+-- lemma erase_top_le (s : composition_series X) : s.erase_top ≤ s :=
+-- begin
+--   rintros x ⟨i, rfl⟩,
+--   simp [mem_def, erase_top],
+-- end
+
+lemma mem_erase_top {s : composition_series X} {x : X}
+  (h : 0 < s.length) : x ∈ s.erase_top ↔ x ≠ s.top ∧ x ∈ s :=
+begin
+  simp only [mem_def],
+  dsimp only [erase_top, coe_fn_mk],
+  split,
+  { rintros ⟨i, rfl⟩,
+    have hi : (i : ℕ) < s.length,
+    { conv_rhs { rw [← nat.succ_sub_one s.length, nat.succ_sub h] },
+      exact i.2 },
+    simp [top, fin.ext_iff, (ne_of_lt hi)] },
+  { intro h,
+    exact mem_erase_top_of_ne_of_mem h.1 h.2 }
+end
+
+lemma lt_top_of_mem_erase_top
+  {s : composition_series X} {x : X}
+  (h : 0 < s.length)
+  (hx : x ∈ s.erase_top) :
+  x < s.top :=
```

Comment:
I don't particularly mind, but this seems excessively many newlines. :-)

Reply:
I agree.